_posts/2016-05-05-cargo-pillars.md

```diff
@@ -0,0 +1,489 @@
+
+Now that we've written our snazzy new `datetime` crate, we'd love to write an
+example to show other developers how it should be used. We create a new file
+called `examples/date.rs` that looks like this:
+
+```
+extern crate datetime;
+
+fn main() {
+    println!("{}", datetime::DateTime::now());
+}
+```
+
+To run the example, we ask Cargo to build and run it:
+
+```
+$ cargo run --example date
+   Compiling datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+     Running `target/debug/examples/date`
+26 Apr 2016 :: 05:03:38
+```
+
+Because we put our code in the conventional location for examples, Cargo knew
+how to do the right thing, no sweat.
+
+In addition, once you start writing a few tests, `cargo test` will automatically
+build your examples as well, which prevents them from getting out of sync with
+your code, and ensures they continue to compile as long as your tests are
+passing.
+
+These are just two examples of a general approach: **Cargo defines a common set of
+conventions and workflows that operate precisely the same way across the entire
+Rust ecosystem**.
+
+### Updating
+
+All of this means that your application won't change if you don't make any
+changes to your dependencies, but what happens when you need to change them?
+
+Cargo adds another layer of protection with **conservative updates**. This means
+that if you modify your `Cargo.toml`, Cargo attempts to minimize the changes
+made to the `Cargo.lock`. The intuition of conservative updates is: **if the
+change you made was unrelated to another dependency, it shouldn't change**.
+
+Let's say that after developing the library for a little while, we decide that
+we want to add support for time zones. First, let's add in the `tz` dependency
+to our package:
+
+```
+  [package]
+  name = "datetime"
+  version = "0.1.0"
+  authors = ["Yehuda Katz <wycats@gmail.com>"]
+
+  [dependencies]
+  time = "0.1.35"
++ tz = "0.2.1"
+```
+
+After using the crate in our library, let's run `cargo build` again:
+
+```
+$ cargo build
+    Updating registry `https://github.com/rust-lang/crates.io-index`
+ Downloading tz v0.2.1
+ Downloading byteorder v0.5.1
+   Compiling byteorder v0.5.1
+   Compiling tz v0.2.1
+   Compiling datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+Cargo downloaded `tz` (and its dependency `byteorder`) and compiled them, but it
+**didn't** touch the packages we were already using (`kernel32-sys`, `libc`,
+`time`, `winapi` and `winapi-build`). Even if one of those package authors
+published an update in the meantime, you can be sure that adding new crates
+won't mess with unrelated ones.
+
+Conservative updates attempt to significantly reduce unexpected changes to your
+source code. It stands in stark contrast to 'rebuild the world', which allows a
+small change to dependencies to rebuild the entire graph, wreaking havoc in its
+wake.
+
+**As a rule, Cargo attempts to minimize the effects of intentional changes to
```

Comment:
This short section has 4 separate bolded things; it's a little distracting, but not too bad.